File: chapter2/expedition.cpp
/* Expedition
 * プライオリティキュー
 * ガソリンスタンドiを通過した時に、プライオリティキューにB_iを追加する
 * 燃料タンクが空になった時
 *  - プライオリティキューが空であれば、移動は達成できない
 *  - プライオリティキューからもっとも大きい数字を取り出し、そこでガソリンを補給した時にする
 */

#include <iostream>
#include <queue>

using namespace std;

int main(){
  int L, P, N;
  cin >> N >> L >> P;
  int A[N + 1], B[N + 1];
  for(int i = 0; i < N; i++){
    cin >> A[i];
  }
  for(int i = 0; i < N; i++){
    cin >> B[i];
  }

  // ゴールをガソリンスタンドを追加
  A[N] = L;
  B[N] = 0;
  N++;

  // ガソリンスタンドを管理する順位キュー
  priority_queue<int> que;

  // ans:補給回数、pos:今いる場所、tank:タンクのガソリンの量
  int ans = 0, pos = 0, tank = P;

  for(int i = 0; i < N; i++){
    // 次に進む距離
    int d = A[i] - pos;

    // 十分な量になるまでガソリンを補給
    while(tank - d < 0){
      if(que.empty()){
        cout << -1 << endl;
        return 0;
      }
      tank += que.top();
      que.pop();
      ans++;
    }
    tank -= d;
    pos = A[i];
    que.push(B[i]);
  }

  cout << ans << endl;

  return 0;
}

File: python/chapter2/expedition.py
"""
トラックで距離Lを移動する。はじめにガソリンがP積まれている。
距離１移動するごとにガソリン１消費する。
途中にはガソリンスタンドがN個あり、i番目のガソリンスタンドはスタートから距離A_iにあり、B_iだけガソリンを補給できる。
最小で何回の補給で移動できるか。移動できない場合は-1を出力する。

4
25
10
10 10
14 5
20 2
21 4
=>2
"""
import queue

n = int(input())
l = int(input())
p = int(input())
g_li = list()

for _ in range(n):
    g_li.append(list(map(int, input().split())))

def solve():
    """
    ガソリンが0になったタイミングで追加した中から
    もっともガソリンの補給量が多いところから補給する

    Returns
    ----------
    ans : int
        最小の補給回数
    """
    que = queue.PriorityQueue()
    ans = 0
    pos = 0
    tank = p

    for i in range(n):
        next_d = g_li[i][0] - pos
        while tank - next_d < 0:
            if (que.empty()):
                print(-1)
                exit()
            tank += que.get()
            ans += 1
        tank -= next_d
        pos = g_li[i][0]
        que.put(g_li[i][1])

    return ans


ans = solve()
print(ans)
